oss_src/unity/extensions/json/decoder.cpp

```diff
@@ -0,0 +1,326 @@
+#include "decoder.hpp"
+#include "types.hpp"
+
+#include <unity/lib/gl_sarray.hpp>
+#include <unity/lib/gl_sframe.hpp>
+#include <unity/lib/gl_sgraph.hpp>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/archive/iterators/binary_from_base64.hpp>
+#include <boost/archive/iterators/transform_width.hpp>
+
+#include <cassert>
+
+using namespace graphlab;
+
+typedef std::map<std::string, variant_type> schema_t;
+
+// forward declarations
+static variant_type _any_from_serializable(const flexible_type& data, const schema_t& schema);
+static schema_t schema_from_variant(const variant_type& v);
+
+static flexible_type _dict_get(const flex_dict& dict, const flex_string& key) {
+  for (const auto& kv : dict) {
+    if (kv.first == key) {
+      return kv.second;
+    }
+  }
+  std::stringstream msg;
+  msg << "Expected key \"";
+  msg << key;
+  msg << "\" was not present in dictionary input.";
+  throw msg.str();
+}
+
+static flexible_type _flexible_type_from_variant(const variant_type& v) {
+  switch (v.which()) {
+    case 0:
+      // flexible_type
+      return variant_get_value<flexible_type>(v);
+    default:
+      throw "Type mismatch: expected flexible_type as input to flex_from_variant.";
+  }
+}
+
+static flex_string _flex_string_from_variant(const variant_type& v) {
+  flexible_type f = _flexible_type_from_variant(v);
```

Comment:
This function can probably be replaced with just a call to
```
variant_get_value<flex_string>(v)
```
It does raise an exception of something like "Expecting blah but we got a bleh" if it does not actually contain a string.